CookSmart/CookSmart/Model/IngredientsStore.swift

```diff
@@ -0,0 +1,230 @@
+    flatIngredients.firstIndex { $0.id == id }!
+  }
+
+  private func flattenedIndexFor(groupIndex: Int, ingredientIndex: Int) -> Int {
+    ingredientGroups.prefix(groupIndex).reduce(0) { $0 + $1.ingredients.count } + ingredientIndex
+  }
+
+  private let ingredientGroupsSourceLock = NSRecursiveLock()
+  private func withIngredientGroupsSourceLock<T>(_ closure: () -> T) -> T {
+    ingredientGroupsSourceLock.lock()
+    defer { ingredientGroupsSourceLock.unlock() }
+    return closure()
+  }
+
+  private static func copyIngredientsFromBundle() {
+    do {
+      try FileManager.default.copyItem(at: bundleIngredientsURL, to: diskIngredientsURL)
+    } catch {
+      logAndFailAssertion("Unable to copy the ingredients plist from the app bundle", error)
+    }
+  }
+
+  private static var diskIngredientsURL: URL {
+    guard let documentDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
+      logAndFailAssertion("Unable to get the path to the documents directory")
+    }
+    if #available(iOS 16.0, *) {
+      return documentDir.appending(path: "ingredients.plist")
+    } else {
+      return documentDir.appendingPathComponent("ingredients.plist")
+    }
+  }
+
+  private static var bundleIngredientsURL: URL {
+    guard let url = Bundle.main.url(forResource: "Ingredients", withExtension: "plist") else {
+      logAndFailAssertion("Unable to find the ingredients plist in the app bundle")
+    }
+    return url
+  }
+
+  // MARK: ObjC Interop Cruft
+
+  @objc func markAccessOfIngredientAtFlattenedIndex(_ flattenedIndex: Array.Index) {
+    withIngredientGroupsSourceLock {
+      if let existingIngredient = self[safeFlattenedIndex: flattenedIndex] {
+        self[existingIngredient.id] = Ingredient(
+          id: existingIngredient.id,
+          name: existingIngredient.name,
+          density: existingIngredient.density,
+          lastAccessDate: Date()
+        )
+      }
+    }
+  }
+
+  @objc func ingredientNameAtFlattenedIndex(_ flattenedIndex: Int) -> String? {
+    self[safeFlattenedIndex: flattenedIndex]?.name
+  }
+
+  @objc func flattenedCountOfIngredients() -> Int {
+    flatIngredients.count
+  }
+
+  @objc func ingredientAnalyticsDictForIngredientAtFlattenedIndex(_ flattenedIndex: Int) -> [String: Any] {
+    let ingr = flatIngredients[flattenedIndex]
+    return [
+      "ingredient_name": ingr.name,
+      "ingredient_density": ingr.density.analyticsRepresentation,
+      "ingredient_id": ingr.id.uuidString,
+    ]
+  }
+
+  @objc func flattenedIndexForGroupIndex(_ groupIndex: Int, ingredientIndex: Int) -> Int {
+    flattenedIndexFor(groupIndex: groupIndex, ingredientIndex: ingredientIndex)
+  }
+
+  @objc func ingredientDictAtFlattenedIndex(_ flattenedIndex: Int) -> [String: Any] {
+    self[flattenedIndex].toDictionary()
+  }
+}
+
+struct IngredientRef {
+  let storedIngredientGroupIndex: Int
+  let ingredientIndex: Int
+}
```

Comment:
wow all this to add your own ingredients? do we have logging for how many people actually use that feature? idk how useful it is.